ngraph/core/reference/include/ngraph/runtime/reference/proposal.hpp

```diff
@@ -0,0 +1,453 @@
+                                const T pred_h = std::exp(d_log_h) * static_cast<T>(hh);
+
+                                // update upper-left corner location
+                                x0 = pred_ctr_x - 0.5f * pred_w;
+                                y0 = pred_ctr_y - 0.5f * pred_h;
+                                // update lower-right corner location
+                                x1 = pred_ctr_x + 0.5f * pred_w;
+                                y1 = pred_ctr_y + 0.5f * pred_h;
+
+                                // adjust new corner locations to be within the image region,
+                                if (clip_before_nms)
+                                {
+                                    x0 = std::max<float>(
+                                        0.0f, std::min<float>(x0, img_W - coordinates_offset));
+                                    y0 = std::max<float>(
+                                        0.0f, std::min<float>(y0, img_H - coordinates_offset));
+                                    x1 = std::max<float>(
+                                        0.0f, std::min<float>(x1, img_W - coordinates_offset));
+                                    y1 = std::max<float>(
+                                        0.0f, std::min<float>(y1, img_H - coordinates_offset));
+                                }
+
+                                // recompute new width & height
+                                const float box_w = x1 - x0 + coordinates_offset;
+                                const float box_h = y1 - y0 + coordinates_offset;
+
+                                proposals[proposal_off + anchor].x0 = static_cast<T>(x0);
+                                proposals[proposal_off + anchor].y0 = static_cast<T>(y0);
+                                proposals[proposal_off + anchor].x1 = static_cast<T>(x1);
+                                proposals[proposal_off + anchor].y1 = static_cast<T>(y1);
+                                proposals[proposal_off + anchor].score =
+                                    static_cast<T>((min_box_W <= box_w) * (min_box_H <= box_h)) *
+                                    score;
+                            }
+                        }
+                    }
+                }
+
+                template <typename T>
+                static void nms(const int num_boxes,
+                                std::vector<int>& is_dead,
+                                // it was
+                                // const std::vector<T>& boxes,
+                                const std::vector<ProposalBox<T>>& proposals,
+                                std::vector<int>& index_out,
+                                int* const num_out,
+                                const int base_index,
+                                const float nms_thresh,
+                                const int max_num_out,
+                                T coordinates_offset)
+                {
+                    int count = 0;
```

Comment:
Looks like we don't need to have this local variable if we increase `num_out` inside the loop.